File: src/block/extrinsics/disputes.rs
// src/extrinsics/disputes.rs

use crate::constants;
use crate::types::{Ed25519PublicKey, Ed25519Signature, ValidatorIndex};

#[derive(Debug, Clone)]
pub struct Judgement {
    pub valid: bool,
    pub validator_index: ValidatorIndex,
    pub signature: Ed25519Signature,
}

#[derive(Debug, Clone)]
pub struct Verdict {
    pub work_report_hash: [u8; 32],
    pub epoch_index: u64, // Note: In your comment, you mention this must be an epoch index of the prior state.
    // We assume that constants::NUM_VALIDATOR_SAFETY_THRESHOLD is defined as a usize.
    pub judgements: [Judgement; constants::NUM_VALIDATOR_SAFETY_THRESHOLD],
}

#[derive(Debug, Clone)]
pub struct Culprit {
    pub invalid_work_report_hash: [u8; 32],
    pub validator_key: Ed25519PublicKey,
    pub signature: Ed25519Signature,
}

#[derive(Debug, Clone)]
pub struct Fault {
    pub work_report_hash: [u8; 32],
    pub incorrect_validity: bool,
    pub validator_key: Ed25519PublicKey,
    pub signature: Ed25519Signature,
}

#[derive(Debug, Clone)]
pub struct Disputes {
    pub verdicts: Vec<Verdict>,
    pub culprits: Vec<Culprit>,
    pub faults: Vec<Fault>,
}

File: go/block/extrinsics/disputes.go
package extrinsics

import (
	"github.com/ascrivener/jam/constants"
	"github.com/ascrivener/jam/types"
)

// todo: validate disputes
type Disputes struct {
	Verdicts []Verdict // v
	Culprits []Culprit // c
	Faults   []Fault   // f
}

func (d Disputes) ToSumOfValidJudgementsMap() map[[32]byte]int {
	m := make(map[[32]byte]int)
	for _, verdict := range d.Verdicts {
		m[verdict.WorkReportHash] = verdict.SumOfValidJudgements()
	}
	return m
}

type Verdict struct {
	WorkReportHash [32]byte                                         // r
	EpochIndex     uint32                                           // a
	Judgements     [constants.NumValidatorSafetyThreshold]Judgement // j
}

func (v Verdict) SumOfValidJudgements() int {
	sum := 0
	for _, judgement := range v.Judgements {
		if judgement.Valid {
			sum++
		}
	}
	return sum
}

type Judgement struct {
	Valid          bool                   // v
	ValidatorIndex types.ValidatorIndex   // i
	Signature      types.Ed25519Signature // s
}

type Culprit struct {
	InvalidWorkReportHash [32]byte               // r
	ValidatorKey          types.Ed25519PublicKey // k
	Signature             types.Ed25519Signature // s
}

type Fault struct {
	WorkReportHash  [32]byte               // r
	CorrectValidity bool                   // v
	ValidatorKey    types.Ed25519PublicKey // k
	Signature       types.Ed25519Signature // s
}
